server/py/framework/db/sqldb/db.py

```diff
@@ -331,6 +331,53 @@ def update_run(self, session, updates: dict, uid: str, project: str, iter: int =
         self._delete_empty_labels(session, Run.Label)
         return run.struct
 
+    def set_run_retrying_status(
+        self,
+        session: Session,
+        project: str,
+        uid: str,
+        retrying: bool,
+    ) -> dict:
+        """
+        Atomically acquire a FOR UPDATE lock on the specified run row, then add or remove
+        the `retrying` label.
+
+        :param session:  SQLAlchemy session to use for the transaction.
+        :param project:     Name of the project containing the run.
+        :param uid:      UID of the workflow‐runner run to lock and update.
+        :param retrying:    Whether to mark the run as retrying (True) or clear that flag (False).
+                            - When setting to True, this will:
+                              1. lock the row
+                              2. verify no existing `retrying` label (else MLRunConflictError)
+                              3. add `retrying="true"`
+                            - When setting to False, it will remove the `retrying` label.
+        :returns:           The updated struct of the run.
+        :raises MLRunNotFoundError:   If the run does not exist.
+        :raises MLRunConflictError:   If attempting to set `retrying=True` when already marked.
+        """
+        run = self._get_run(session, uid, project, iteration=0, with_for_update=True)
+        if not run:
+            raise mlrun.errors.MLRunNotFoundError(f"Run {project}/{uid} not found")
+
+        struct = run.struct
+        labels = run_labels(struct)
+
+        if not retrying:
+            labels.pop("retrying", None)
+        elif mlrun_constants.MLRunInternalLabels.retrying in labels:
+            # flush and commit so the lock is released immediately
+            session.commit()
```

Comment:
I understand the idea of committing to release the lock but why I see you dont do it on line 359 above. perhaps a mistake? anyway, perhaps I think if the commit needs to happen on any exception being raised, Id wrap everything in try/except and handle the commit upon exception.